### PR TITLE
sqldb+kvdb: unify SQL error mapping and transaction retry for between the `kvdb` and `sqldb` packages

### PR DESCRIPTION
**Pull request by bhandras**
Previously we had very similar error mapping and retry functions in the affected modules diverging over time. To avoid bugs resulting from individual updates this PR consolidates the retry and error mapping to live in the newly carved out `sqldb` package.

Based on: https://github.com/lightningnetwork/lnd/pull/8603
Added commits from : https://github.com/lightningnetwork/lnd/pull/8529

### CODE REVIEW

File: sqldb/interfaces.go [hunk 1]
```diff
@@ -82,6 +88,12 @@ type BatchedQuerier interface {
 	BeginTx(ctx context.Context, options TxOptions) (*sql.Tx, error)
 }
 
+type SqlcBatchedQuerier interface {
```
Comment (guggero) [inline sqldb/interfaces.go:91]: nit: missing Godoc.

File: kvdb/sqlbase/db.go [hunk 2]
```diff
@@ -227,112 +215,43 @@ func (db *db) Update(f func(tx walletdb.ReadWriteTx) error,
...
-			}
-
-			return dbErr
+	execTxBody := func(tx sqldb.Tx) error {
+		kvTx, ok := tx.(*readWriteTx)
+		if !ok {
+			return fmt.Errorf("expected *readWriteTx, got %T", tx)
 		}
 
-		err = catchPanic(func() error { return f(tx) })
```
Comment (guggero) [inline kvdb/sqlbase/db.go:311]: I wonder if we want to port the `catchPanic` over to the common code as well? Since apparently the wallet sometimes panics? Or why was this added in the first place?

File: go.mod [hunk 3]
```diff
@@ -206,6 +206,9 @@ replace google.golang.org/protobuf => github.com/lightninglabs/protobuf-go-hex-d
 // Temporary replace rule until sqldb is tagged.
 replace github.com/lightningnetwork/lnd/sqldb => ./sqldb
 
+// Temporary replace rule until kvdb is tagged.
+replace github.com/lightningnetwork/lnd/kvdvb => ./kvdb
```
Comment (guggero) [inline go.mod:210]: This seems to be misspelled... Weird that nothing seems to break because of it? Or maybe the replace doesn't take effect after all and we're still running with the old code?

File: kvdb/sqlbase/db.go (hunk 2 above)
Comment (guggero) [reply kvdb/sqlbase/db.go:311]: Oops, never mind, I see that it's part of the `execTxBody` now...

File: channeldb/payment_control.go [hunk 4]
```diff
@@ -229,7 +229,7 @@ func (p *PaymentControl) InitPayment(paymentHash lntypes.Hash,
 		return bucket.Delete(paymentFailInfoKey)
 	})
 	if err != nil {
-		return err
+		return fmt.Errorf("unable to init payment: %w", err)
```
Comment (guggero) [inline channeldb/payment_control.go:232]: This doesn't seem to match the commit message anymore... Did we find out the cause for the wrapping to cause issues and changed it back?

**Review by guggero — COMMENTED** [review]
Looks very good! Just a few minor things, I think we can get this merged pretty soon :tada:

File: channeldb/payment_control.go (hunk 4 above)
Comment (bhandras) [reply channeldb/payment_control.go:232]: Yeah I think the other PR had some incremental changes that were not cleaned up so this commit doesn't make sense anymore. I'm squashing it with the "wrap all errors" one.

Comment (Roasbeef) [reply channeldb/payment_control.go:232]: IIRC we had to _revert_ this part to have the error introspection work properly: https://github.com/lightningnetwork/lnd/pull/8529#discussion_r1516997708

File: sqldb/interfaces.go [hunk 5]
```diff
@@ -334,6 +334,8 @@ func (t *TransactionExecutor[Q]) ExecTx(ctx context.Context,
 		if !ok {
 			return fmt.Errorf("expected *sql.Tx, got %T", tx)
 		}
+
+		reset()
```
Comment (Roasbeef) [inline sqldb/interfaces.go:338]: I wonder if missing this partially exacerbated the issues w/ the async payment benchmark?

**Comment by guggero** [comment]
Can be rebased. And the module `sqldb/v1.0.0` tag can now be used instead of the local replace.

File: kvdb/sqlbase/db.go [hunk 6]
```diff
@@ -169,15 +169,25 @@ func (db *db) getPrefixedTableName(table string) string {
...
 			default:
 				err = errors.New(fmt.Sprintf("%v", data))
 			}
+
+			// Before we issue a critical log which'll cause the
+			// daemon to shut down, we'll first check if this is a
+			// DB serialization error. If so, then we don't need to
+			// log as we can retry safely and avoid tearing
+			// everything down.
+			if sqldb.IsSerializationError(sqldb.MapSQLError(err)) {
```
[the remark this answers is not in the document]
Comment (bhandras) [reply kvdb/sqlbase/db.go:185]: This panic handler remains in `kvdb` as it is indeed specific to how the postgres impl works. We'll just use the template `sqldb.ExecuteSQLTransactionWithRetry` later. So the change is all about removing the critical log (taken from the lock PR).

File: invoices/sql_store.go [hunk 7]
```diff
@@ -259,7 +259,7 @@ func (i *SQLStore) AddInvoice(ctx context.Context,
 			AddedAt:   newInvoice.CreationDate.UTC(),
 			InvoiceID: invoiceID,
 		})
-	})
+	}, func() {})
```
[the remark this answers is not in the document]
Comment (bhandras) [reply invoices/sql_store.go:262]: done, ptal

File: invoices/sql_store.go [hunk 8]
```diff
@@ -683,6 +684,7 @@ func (i *SQLStore) InvoicesSettledSince(ctx context.Context, idx uint64) (
 				SettleIndexGet: sqldb.SQLInt64(settleIdx + 1),
 				NumLimit:       int32(limit),
 				NumOffset:      int32(offset),
+				Reverse:        false,
```
[the remark this answers is not in the document]
Comment (bhandras) [reply invoices/sql_store.go:687]: Yeah, it was pretty hard to catch as the flake is very rare. Initially I thought that the variable reset (see previous commit) was the root cause, but turns out our query using named params make the reverse flag nullable.

File: sqldb/sqlc/queries/invoices.sql [hunk 9]
```diff
@@ -78,14 +78,14 @@ WHERE (
     END
 )
 ORDER BY
-    CASE
-        WHEN sqlc.narg('reverse') = FALSE THEN id  
-        ELSE NULL
+CASE
+    WHEN sqlc.narg('reverse')=FALSE OR sqlc.narg('reverse') IS NULL THEN id
```
[the remark this answers is not in the document]
Comment (bhandras) [reply sqldb/sqlc/queries/invoices.sql]: done

File: sqldb/sqlc/queries/invoices.sql [hunk 10]
```diff
@@ -78,14 +78,14 @@ WHERE (
...
+    ELSE NULL
     END ASC,
-    CASE
-        WHEN sqlc.narg('reverse') = TRUE  THEN id  
-        ELSE NULL
-    END DESC
+CASE
+    WHEN sqlc.narg('reverse')=TRUE THEN id
```
[the remark this answers is not in the document]
Comment (bhandras) [reply sqldb/sqlc/queries/invoices.sql]: done

**Comment by bhandras** [comment]
> Can be rebased. And the module `sqldb/v1.0.0` tag can now be used instead of the local replace.

Can't fully use the tag yet as this PR also changes both `sqldb` and `kvdb`. We can tag and update in a follow up though.

**Comment by bhandras** [comment]
@coderabbitai pause

File: go.mod [hunk 11]
```diff
@@ -203,7 +203,12 @@ replace github.com/gogo/protobuf => github.com/gogo/protobuf v1.3.2
 // allows us to specify that as an option.
 replace google.golang.org/protobuf => github.com/lightninglabs/protobuf-go-hex-display v1.30.0-hex-display
 
-// Temporary replace rule until sqldb is tagged.
+// Temporary replace rule until kvdb is tagged with changes from
+// https://github.com/lightningnetwork/lnd/pull/8611.
+replace github.com/lightningnetwork/lnd/kvdb => ./kvdb
+
+// Temporary replace rule until sqldb is tagged with changes from
+// https://github.com/lightningnetwork/lnd/pull/8611.
```
Comment (Roasbeef) [inline go.mod:211]: Should be able to remove both this and the one above now.

File: kvdb/go.mod [hunk 12]
```diff
@@ -113,4 +137,6 @@ replace github.com/ulikunitz/xz => github.com/ulikunitz/xz v0.5.11
 // https://deps.dev/advisory/OSV/GO-2021-0053?from=%2Fgo%2Fgithub.com%252Fgogo%252Fprotobuf%2Fv1.3.1
 replace github.com/gogo/protobuf => github.com/gogo/protobuf v1.3.2
 
-go 1.19
+replace github.com/lightningnetwork/lnd/sqldb => ../sqldb
```
Comment (Roasbeef) [inline kvdb/go.mod:140]: Same here re removing the replace.

File: kvdb/go.mod [hunk 13]
```diff
@@ -113,4 +137,6 @@ replace github.com/ulikunitz/xz => github.com/ulikunitz/xz v0.5.11
...
+
+go 1.21.4
```
Comment (Roasbeef) [inline kvdb/go.mod:142]: I think we want to keep this at Go 1.19 as we hadn't yet bumped the Go versions repo wide yet.

File: kvdb/go.mod (hunk 12 above)
Comment (bhandras) [reply kvdb/go.mod:140]: I think this is also needed as the PR modifies `sqldb` and is used in `kvdb`.

File: kvdb/go.mod (hunk 13 above)
Comment (bhandras) [reply kvdb/go.mod:144]: This is bumped because `kvdb` now depends on `sqldb` which is 1.21.4. On the other hand I think we're safe to bump as the main go.mod also is using 1.21.4.

File: kvdb/go.mod (hunk 12 above)
Comment (bhandras) [reply kvdb/go.mod:142]: Added a comment to clarify.

**Review by Roasbeef — APPROVED** [review]
LGTM 🪡
